### PR TITLE
dcnm_image_upload module initial code drop

### PR DESCRIPTION
**Pull request by mmudigon**
This commit includes:
  - Code for uploading/deleting/querying images from controller
  - Unit test cases and the corresponding code
  - IT cases

Tracking Issue: https://github.com/CiscoDevNet/ansible-dcnm/issues/258

### CODE REVIEW

File: plugins/httpapi/dcnm.py [hunk 1]
```diff
@@ -62,6 +62,12 @@ def get_version(self):
     def set_version(self, version):
         self.version = version
 
+    def set_token(self, token):
+        self.token = token
```
Comment (mikewiebe) [inline plugins/httpapi/dcnm.py:66]: Why is this needed?  Does this have any impact on other modules in the collection?  Did we run tests to confirm?

File: plugins/module_utils/network/dcnm/dcnm.py [hunk 2]
```diff
@@ -23,6 +23,13 @@
 from ansible.module_utils.common import validation
 from ansible.module_utils.connection import Connection
 
+# Any third party module must be imported as shown. If not ansible sanity tests will fail
+try:
```
Comment (mikewiebe) [inline plugins/module_utils/network/dcnm/dcnm.py:27]: Can we add an import skip instead similar to this?

https://www.cisco.com/c/dam/en/us/td/docs/dcn/ndfc/1213/articles/ndfc-vpc-fabric/configuring-vpc-fabric-peering.pdf

File: plugins/module_utils/network/dcnm/dcnm.py [hunk 3]
```diff
@@ -386,7 +393,7 @@ def dcnm_reset_connection(module):
 
     conn = Connection(module._socket_path)
 
-    conn.logout()
+    # conn.logout()
```
Comment (mikewiebe) [inline plugins/module_utils/network/dcnm/dcnm.py:396]: Remove?

File: plugins/module_utils/network/dcnm/dcnm.py [hunk 4]
```diff
@@ -517,3 +524,32 @@ def dcnm_get_url(module, fabric, path, items, module_name):
         iter += 1
 
     return attach_objects
+
+
+def dcnm_get_protocol_and_address(module):
```
Comment (mikewiebe) [inline plugins/module_utils/network/dcnm/dcnm.py:529]: Can you add some comments on the purpose of these 3 new functions?  Also why do we need the new dcnm_post_request method?

File: plugins/modules/dcnm_image_upload.py [hunk 5]
```diff
@@ -0,0 +1,777 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2020-2022 Cisco and/or its affiliates.
```
Comment (mikewiebe) [inline plugins/modules/dcnm_image_upload.py:3]: Update Copyright

File: tests/unit/modules/dcnm/test_dcnm_image_upload.py [hunk 6]
```diff
@@ -0,0 +1,865 @@
+# Copyright (c) 2020-2022 Cisco and/or its affiliates.
```
Comment (mikewiebe) [inline tests/unit/modules/dcnm/test_dcnm_image_upload.py:1]: Update Copyright

File: plugins/modules/dcnm_image_upload.py [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+        Returns:
+            True if file successfully transfered
+            False otherwise
+        """
+
+        protocol, address = dcnm_get_protocol_and_address(self.module)
+
+        path = protocol + ":" + address + self.paths["DCNM_CREATE_IMAGE_LOCAL"]
+
+        # We should use the authentication token already obtained to send data over this connection.
```
Comment (mikewiebe) [inline plugins/modules/dcnm_image_upload.py:629]: Is this something we need to fix in our dcnm_send function?

File: plugins/httpapi/dcnm.py (hunk 1 above)
Comment (mikewiebe) [reply plugins/httpapi/dcnm.py:66]: Please add a comment for why this is needed as discussed during our call

File: plugins/module_utils/network/dcnm/dcnm.py (hunk 2 above)
Comment (mikewiebe) [reply plugins/module_utils/network/dcnm/dcnm.py:27]: Discussed this and we can keep it as is

File: plugins/httpapi/dcnm.py (hunk 1 above)
Comment (mmudigon) [reply plugins/httpapi/dcnm.py:66]: Sure, will do

File: plugins/module_utils/network/dcnm/dcnm.py (hunk 3 above)
Comment (mmudigon) [reply plugins/module_utils/network/dcnm/dcnm.py:396]: done

File: plugins/modules/dcnm_image_upload.py (hunk 5 above)
Comment (mmudigon) [reply plugins/modules/dcnm_image_upload.py:3]: ok

File: tests/unit/modules/dcnm/test_dcnm_image_upload.py (hunk 6 above)
Comment (mmudigon) [reply tests/unit/modules/dcnm/test_dcnm_image_upload.py:1]: will do

File: plugins/modules/dcnm_image_upload.py (hunk 7 above)
Comment (mmudigon) [reply plugins/modules/dcnm_image_upload.py:629]: As explained in the meeting, we are using "requests" module to send multi-part-frames for uploading images. For using requests we should build auth headers and also the complete URL appropriately. So we cannot use dcnm_send here.

File: plugins/module_utils/network/dcnm/dcnm.py (hunk 4 above)
Comment (mmudigon) [reply plugins/module_utils/network/dcnm/dcnm.py:529]: ok
